core/src/test/scala/unit/kafka/coordinator/transaction/TransactionMarkerChannelManagerTest.scala

```diff
@@ -86,6 +90,70 @@ class TransactionMarkerChannelManagerTest {
       .anyTimes()
   }
 
+  @Test
+  def shouldOnlyWriteTxnCompletionOnce(): Unit = {
```

Comment:
Does this test cover concurrent calls to `maybeWriteTxnCompletion`?

Reply:
It does. I was trying to setup this test to fit how we're likely hitting this in practice. In the call to `addTxnMarkersToSend`, before calling `maybeWriteTxnCompletion`, we have to acquire the lock. It is possible that the caller fails to acquire the lock before the markers finish getting written and the transaction gets completed in the request completion handler.

Reply:
Got it, so when the bug still exist this test would probably not fail consistently, but would be flaky, right?

Reply:
It fails deterministically without the fix.

Reply:
Ah yes, I missed the `txnMetadata2.lock.lock()` before starting the scheduler, thanks.